### ISSUE TITLE
type-narrowing causes issues with TypeVars later in function body

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#2011]
**Describe the bug**
type-narrowing appends extra union types to a TypeVar outside the narrowed scope, causing issues later down the line.

**To Reproduce**
See code below.

Pyright returns:
```
test.py:7:17 - info: Type of "arg" is "T@foo"
test.py:10:21 - info: Type of "arg" is "str"
test.py:12:17 - info: Type of "arg" is "str | T@foo"
test.py:13:12 - error: Expression of type "str | T@foo" cannot be assigned to return type "T@foo"
    Type "str | T@foo" cannot be assigned to type "T@foo" (reportGeneralTypeIssues)
```

**Expected behavior**
Outside the `if isinstance` block, `arg`'s type would be `T` alone, not a union of `str` and `T`.

**Screenshots or Code**
Consider the following:
```python
from typing import TypeVar


T = TypeVar("T")


def foo(arg: T) -> T:
    reveal_type(arg)

    if isinstance(arg, str):
        reveal_type(arg)

    reveal_type(arg)
    return arg

```
![image](https://user-images.githubusercontent.com/8677174/122628961-c88dbc00-d0fc-11eb-9350-e542415b65f3.png)

**VS Code extension or command-line**
pyright 1.1.150

**Additional context**
Mypy has no issues.

**Comment by erictraut** [comment]
This is an interesting one. Thanks for reporting it.

Type narrowing is not meant to apply only to the narrowed scope. Narrowed types are "merged" based on the code flow graph. This is important in many cases. For example:

```python
def foo(arg: object):
def verify_str_or_bytes(arg: object) -> str | bytes:
    if not isinstance(arg, (str, bytes)):
        raise Exception("Not AnyStr")
    reveal_type(arg) # str | bytes
    return arg
```

The challenge with your code sample is that the `isinstance` type narrowing is "forgetting" that the narrowed type has any relation to the original type `T`.

Mypy must have some logic to handle this very specific case. If I modify the code only slightly, mypy demonstrates different (inconsistent) behaviors. This is indicative of an attempt to plug a specific hole without covering the general case.

```python
def foo(arg: T) -> T:
    # Flip the condition from `isinstance` to `not isinstance`
    if not isinstance(arg, str):
        reveal_type(arg) # mypy says: "T`-1"
        return arg

    reveal_type(arg) # mypy says: "builtins.str"
    return arg # mypy says: "error: Incompatible return value type"
```

```python
def foo(arg: T) -> T:
    if isinstance(arg, str):
        reveal_type(arg) # mypy says: "builtins.str"
    else:
        # Return if it's not a str
        return arg

    reveal_type(arg) # mypy says: "builtins.str"
    return arg # mypy says: "error: Incompatible return value type"
```

```python
def foo(arg: T) -> T:
    if isinstance(arg, str):
        reveal_type(arg) # mypy says: "builtins.str"
    elif isinstance(arg, int):
        reveal_type(arg) # mypy says: "builtins.int"
    else:
        # Return if it's not a str or an int
        return arg

    reveal_type(arg) # mypy says: "T`-1"
    return arg # mypy says: no error
```

So mypy appears to have some bugs in this general area.

I'll need to think about this more to see if I can come up with an elegant solution.

**Comment by erictraut** [comment]
I found a general (and elegant) solution to this problem. The fix will be in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.155, which I just published. It will also be included in the next release of pylance.
